File: wasm/verify/opacity_test.mjs
// core.Opacity against the live DOM.
//
// htmlout writes the same declaration into a static document and is tested
// there. This is the live half, and it has the two cases a static export
// cannot: the sentinel arriving in a patch, and a node that stops being faded
// having its old declaration removed from the reused element.
//
// The sentinel is the point of the first. An alpha of zero crosses the wire
// as core.OpacityClear (-1), because a plain zero is what an unset field looks
// like. A runtime that wrote the number through would hand CSS "-1", which it
// clamps to 0 and so looks right, until the day the sentinel is read as
// "falsy means unset" and a faded-out node is drawn opaque. So the string is
// pinned, not the pixels. wasm/verify's opacity_test.go holds the number
// itself to core.OpacityClear.

import test from "node:test";
import assert from "node:assert/strict";

import { loadRuntime, nodeAt } from "./load.mjs";

function mount(children) {
    const rt = loadRuntime();
    rt.GrMob.mount(JSON.stringify({ Type: "Column", Children: children }));
    rt.drainFrames();
    return { rt, at: (i) => nodeAt(rt.document, `root/${i}`) };
}

const box = (style) => ({ Type: "Box", Style: style });

test("a faded node gets its alpha, and the sentinel is written as zero", () => {
    const { at } = mount([box({ Opacity: 0.4 }), box({ Opacity: 1 }), box({ Opacity: -1 })]);
    assert.equal(at(0).style.opacity, "0.4");
    assert.equal(at(1).style.opacity, "1");
    assert.equal(at(2).style.opacity, "0",
        "core.OpacityClear must reach the page as an alpha of zero");
});

test("an undeclared opacity writes nothing", () => {
    const { at } = mount([box({})]);
    assert.equal(at(0).style.opacity, "");
});

test("fading out and back in over patches, ending with no declaration", () => {
    // The patch path reuses the element. A node that fades back in by
    // dropping the prop sends a Style with no Opacity, and a guarded write
    // would leave it transparent for good.
    const { rt, at } = mount([box({ Opacity: 0.4, Transition: "200ms ease" })]);
    assert.equal(at(0).style.opacity, "0.4");

    rt.GrMob.patch(JSON.stringify([
        { Type: "update-style", TargetID: "root/0", Changes: { Opacity: -1, Transition: "200ms ease" } },
    ]));
    rt.drainFrames();
    assert.equal(at(0).style.opacity, "0");

    rt.GrMob.patch(JSON.stringify([
        { Type: "update-style", TargetID: "root/0", Changes: { Transition: "200ms ease" } },
    ]));
    rt.drainFrames();
    assert.equal(at(0).style.opacity, "",
        "the node stayed transparent after its Opacity was dropped");
});

File: wasm/verify/opacity_test.go
package main

import (
	"os"
	"regexp"
	"strconv"
	"testing"

	"github.com/rohanthewiz/grmob/core"
)

// core.OpacityClear, held to the three places outside Go that spell it.
//
// # The second sentinel, and why it gets the first one's test
//
// Opacity is the second core.Style number whose CSS initial value is not zero
// (see core.OpacityClear), so an alpha of zero crosses the wire as -1 and each
// runtime needs one rule to read it:
//
//	wasm/grmob-runtime.js     styleFromGrMob, which writes opacity
//	android/.../GrMobStyle.kt alphaOf, resolved at parse
//	ios/.../GrMobStyle.swift  alpha, read by grMobBox
//
// shrink_test.go says why a drifting sentinel is the worst kind: everything
// keeps compiling and each copy starts meaning something different. Here the
// failure is quieter still. CSS clamps a negative opacity to 0, so the web
// would go on looking right with the wrong number while Compose (which throws
// on an alpha outside [0, 1]) and SwiftUI did not.
//
// Each pattern spans the unset arm as well as the sentinel arm, for the reason
// kotlinShrinkSentinel gives: an unset Opacity must read as 1. A runtime that
// defaulted it to the zero it is stored as would draw every node in every
// tree transparent, and no comparison of sentinels would notice.
//
// What mobile/verify holds is the other half of each native mapping: that the
// reading reaches a modifier, at the right layer.

// runtimeOpacitySentinel matches the comparison in the runtime's style mapping.
var runtimeOpacitySentinel = regexp.MustCompile(`style\.Opacity === (-?\d+)\s*\? "0"`)

// kotlinOpacitySentinel matches both arms of GrMobStyle.kt's alphaOf.
var kotlinOpacitySentinel = regexp.MustCompile(
	`(?s)when \(opacity\) \{\s*0f -> 1f\s*(-?\d+)f -> 0f`)

// swiftOpacitySentinel matches both arms of GrMobStyle.swift's alpha.
var swiftOpacitySentinel = regexp.MustCompile(
	`(?s)if opacity == 0 \{ return 1 \}\s*if opacity == (-?\d+) \{ return 0 \}`)

func TestTheOpacitySentinelIsTheSameNumberEverywhere(t *testing.T) {
	want := strconv.Itoa(core.OpacityClear)

	for _, c := range []struct {
		file string
		re   *regexp.Regexp
		what string
	}{
		{"../grmob-runtime.js", runtimeOpacitySentinel,
			"this is the only place the live runtime knows that an alpha of zero is " +
				"not a zero. Reading the wrong number writes the sentinel through as an " +
				"alpha, or reads a real fade-out as \"no declaration\" and draws it opaque"},
		{"../../android/app/src/main/java/com/grmob/runtime/GrMobStyle.kt",
			kotlinOpacitySentinel,
			"alphaOf is the whole of what the Compose runtime knows about the sentinel, " +
				"and Modifier.alpha throws outside [0, 1] where CSS would clamp, so a " +
				"sentinel read as a plain number is a crash and not a wrong pixel"},
		{"../../ios/GrMob/Runtime/GrMobStyle.swift", swiftOpacitySentinel,
			"alpha is what grMobBox hands .opacity, and its unset arm is what keeps " +
				"every node that never mentioned Opacity drawn at all"},
	} {
		raw, err := os.ReadFile(c.file)
		if err != nil {
			t.Fatalf("reading %s: %v", c.file, err)
		}
		m := c.re.FindStringSubmatch(string(raw))
		if m == nil {
			t.Errorf("%s no longer spells the opacity sentinel where this test looks "+
				"for it (%s).\n\n%s\n\nIf the mapping moved, re-point this; if it was "+
				"removed, core.Opacity(0) no longer fades anything out on this target.",
				c.file, c.re, c.what)
			continue
		}
		if m[1] != want {
			t.Errorf("%s spells the opacity sentinel %s and core.OpacityClear is %s.\n\n%s",
				c.file, m[1], want, c.what)
		}
	}
}
